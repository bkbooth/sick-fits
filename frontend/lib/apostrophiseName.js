const ESS = 's'

export default function apostrophiseName(name) {
  const nameEndsWithEss = name[name.length - 1].toLowerCase() === ESS
  return `${name}'${nameEndsWithEss ? '' : ESS}`
}
