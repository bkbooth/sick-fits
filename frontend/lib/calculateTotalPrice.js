/**
 * Calculate the total price of a cart by summing cart items' price multiplied by quantity
 *
 * @param {CartItem[]} cart
 * @param
 *
 * @returns {number}
 */
export default function calculateTotalPrice(cart) {
  return cart.reduce((total, cartItem) => {
    if (!cartItem.item) return total;
    return total + cartItem.quantity * cartItem.item.price;
  }, 0);
}

/**
 * @typedef {Object} CartItem
 * @property {number} quantity
 * @property {Object} item
 * @property {number} item.price
 */
