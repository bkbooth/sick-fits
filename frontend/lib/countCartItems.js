/**
 * Count the total number of items in a cart
 *
 * @param {CartItem[]} cart
 *
 * @returns {number}
 */
export default function countCartItems(cart) {
  return cart.reduce((total, cartItem) => total + cartItem.quantity, 0);
}

/**
 * @typedef {Object} CartItem
 * @property {number} quantity
 */
