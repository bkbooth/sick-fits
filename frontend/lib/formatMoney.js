export default function formatMoney(amount) {
  return new Intl.NumberFormat('en-AU', {
    style: 'currency',
    currency: 'AUD',
    // if its a whole, dollar amount, leave off the .00
    minimumFractionDigits: amount % 100 === 0 ? 0 : 2,
  }).format(amount / 100);
}
