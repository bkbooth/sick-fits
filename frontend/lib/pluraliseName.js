const ESS = 's'

export default function pluraliseName(name) {
  const nameEndsWithEss = name[name.length - 1].toLowerCase() === ESS
  return `${name}'${nameEndsWithEss ? '' : ESS}`
}
