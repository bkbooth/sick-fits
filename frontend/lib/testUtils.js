import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

export const fakeUser = () => ({
  __typename: 'User',
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
});

export const fakeItem = () => ({
  __typename: 'Item',
  id: casual.uuid,
  title: casual.title,
  description: casual.description,
  price: casual.integer(1000, 30000),
  image: `${casual.word}.jpg`,
  largeImage: `${casual.word}-large.jpg`,
  user: null,
});

export const fakeCartItem = overrides => ({
  __typename: 'CartItem',
  id: casual.uuid,
  quantity: casual.integer(1, 5),
  item: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

export function typeInto(wrapper, { inputType = 'input', name, type = 'text', value }) {
  wrapper
    .find(`${inputType}[name="${name}"]`)
    .simulate('change', { target: { name, type, value } });
}
