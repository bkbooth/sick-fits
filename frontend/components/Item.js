import React from 'react';
import Link from 'next/link';
import PropTypes from 'prop-types';
import formatMoney from 'lib/formatMoney';
import DeleteItem from 'components/DeleteItem';
import ItemStyles from 'components/styles/ItemStyles';
import PriceTag from 'components/styles/PriceTag';
import Title from 'components/styles/Title';

const Item = ({ item }) => (
  <ItemStyles>
    {item.image && <img src={item.image} alt={item.title} />}

    <Title>
      <Link href="/items/[itemId]" as={`/items/${item.id}`}>
        <a>{item.title}</a>
      </Link>
    </Title>
    <PriceTag>{formatMoney(item.price)}</PriceTag>
    <p>{item.description}</p>

    <div className="buttonList">
      <Link href="/items/[itemId]/update" as={`/items/${item.id}/update`}>
        <a>✏️ Edit</a>
      </Link>
      <button>Add to cart</button>
      <DeleteItem itemId={item.id}>Delete</DeleteItem>
    </div>
  </ItemStyles>
);

Item.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    image: PropTypes.string,
  }),
};

export default Item;
