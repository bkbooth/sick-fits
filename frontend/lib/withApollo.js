import ApolloClient, { InMemoryCache } from 'apollo-boost';
import withApollo from 'next-with-apollo';
import { LOCAL_STATE_QUERY } from 'components/Cart';
import { GRAPHQL_URL } from '../config';

function createApolloClient({ headers, initialState }) {
  return new ApolloClient({
    uri: GRAPHQL_URL,
    cache: new InMemoryCache().restore(initialState || {}),
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_parent, _args, { cache }, _info) {
            const { cartOpen } = cache.readQuery({ query: LOCAL_STATE_QUERY });
            const result = { data: { cartOpen: !cartOpen } };
            cache.writeData(result);
            return result;
          },
        },
      },
      defaults: {
        cartOpen: false,
      },
    },
  });
}

export default withApollo(createApolloClient);
