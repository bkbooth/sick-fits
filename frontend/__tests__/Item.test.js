import { shallow } from 'enzyme';
import { fakeItem } from 'lib/testUtils';
import Item from 'components/Item';

const item = fakeItem();

describe('<Item />', () => {
  it('renders and matches the snapshot', () => {
    const wrapper = shallow(<Item item={item} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the title and price tag', () => {
    const wrapper = shallow(<Item item={item} />);
    expect(wrapper.find('PriceTag').text()).toBe('$164.46');
    expect(wrapper.find('Title a').text()).toBe(item.title);
  });

  it('renders the image', () => {
    const wrapper = shallow(<Item item={item} />);
    const image = wrapper.find('img');
    expect(image.prop('src')).toBe(item.image);
    expect(image.prop('alt')).toBe(item.title);
  });

  it('renders the buttoms', () => {
    const wrapper = shallow(<Item item={item} />);
    const buttonList = wrapper.find('.buttonList');
    expect(buttonList.children()).toHaveLength(3);
    expect(buttonList.find('Link').exists()).toBe(true);
    expect(buttonList.find('AddToCart').exists()).toBe(true);
    expect(buttonList.find('DeleteItem').exists()).toBe(true);
  });
});
