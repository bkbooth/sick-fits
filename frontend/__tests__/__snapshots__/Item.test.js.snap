// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item /> renders and matches the snapshot 1`] = `
<ItemStyles__Item>
  <img
    alt="Soluta non"
    src="dicta.jpg"
  />
  <Title>
    <Link
      as="/items/3a430a73-c9e9-4bcc-b46a-41618965ffea"
      href="/items/[itemId]"
      prefetch={true}
    >
      <a>
        Soluta non
      </a>
    </Link>
  </Title>
  <PriceTag>
    $164.46
  </PriceTag>
  <p>
    Enim quia autem et modi tenetur amet modi reprehenderit. Adipisci eius laborum eius aut illo dolorum. Aliquam id sit. Vel ipsa tempora aut sint aperiam occaecati illum animi vel.
  </p>
  <div
    className="buttonList"
  >
    <Link
      as="/items/3a430a73-c9e9-4bcc-b46a-41618965ffea/update"
      href="/items/[itemId]/update"
      prefetch={true}
    >
      <a>
        ✏️ Edit
      </a>
    </Link>
    <AddToCart
      itemId="3a430a73-c9e9-4bcc-b46a-41618965ffea"
    />
    <DeleteItem
      itemId="3a430a73-c9e9-4bcc-b46a-41618965ffea"
    />
  </div>
</ItemStyles__Item>
`;
